roofit/roofitcore/src/RooAbsCollection.cxx

```diff
@@ -691,15 +691,55 @@ Bool_t RooAbsCollection::remove(const RooAbsArg& var, Bool_t , Bool_t matchByNam
 
 
 ////////////////////////////////////////////////////////////////////////////////
-/// Remove each argument in the input list from our list using remove(const RooAbsArg&).
+/// Remove each argument in the input list from our list.
+/// An exact pointer match is required, not just a match by name.
+/// If `matchByNameOnly` is set, items will be looked up by name. In this case, if
+/// the collection also owns the items, it will delete them.
 /// Return kFALSE in case of problems.
 
 Bool_t RooAbsCollection::remove(const RooAbsCollection& list, Bool_t silent, Bool_t matchByNameOnly)
 {
 
   auto oldSize = _list.size();
-  for (auto item : list._list) {
-    remove(*item, silent, matchByNameOnly);
+  std::vector<const RooAbsArg*> markedItems;
+
+  if (matchByNameOnly) {
+
+    // Instead of doing two passes on the list as in remove(RooAbsArg&), we do
+    // everything in one pass, by using side effects of the predicate.
+    auto nameMatchAndMark = [&list, &markedItems](const RooAbsArg* elm) {
+      if( list.contains(*elm) ) {
+        markedItems.push_back(elm);
+        return true;
+      }
+      return false;
+    };
+
+    _list.erase(std::remove_if(_list.begin(), _list.end(), nameMatchAndMark), _list.end());
+
+    std::set<const RooAbsArg*> toBeDeleted(markedItems.begin(), markedItems.end());
+    if (_ownCont) {
+      for (auto arg : toBeDeleted) {
+        delete arg;
+      }
+    }
+  }
+  else {
+    auto argMatchAndMark = [&list, &markedItems](const RooAbsArg* elm) {
+      if( list.containsInstance(*elm) ) {
+        markedItems.push_back(elm);
+        return true;
+      }
+      return false;
+    };
+
+    _list.erase(std::remove_if(_list.begin(), _list.end(), argMatchAndMark), _list.end());
+  }
+
+  if (_hashAssistedFind && oldSize != _list.size()) {
+    for( auto& var : markedItems ) {
+      _hashAssistedFind->erase(var);
+    }
```

Comment:
Indeed, if you have a RooArgList, and two elements with the same name, you might remove the first one from the hash map. The find function will now report that there's no element by that name if the hash map is active. If it's inactive, it will report the element. It's something we have to think about. @guitargeek 
This is anyway a bit shaky, since it will only ever find the first element by that name if you ask for this in a list with duplicates. It doesn't seem to be a big issue in RF so far, as nobody seems to use the hash map with ArgLists and duplicates, but it might lead to surprises.